### DISCUSSION TITLE
Basic client / server on Pico W / windows

### DISCUSSION THREAD

**Discussion by Derfies** [micropython/micropython#13531] [RP2040 / Pico]
Hi all,

Only just getting into microcontrollers but enjoying myself so far. Please forgive the silly questions that will no doubt be aired below.

I'm interested in getting a basic wifi example up and running using Pico W and a windows server, running micropython and python respectively. Pretty simple - I push a button attached to the microcontroller, the server receives a message. However it _absolutely_ has to be sent and received, which led me down the path of resilient client / server libs like [this one:](https://github.com/peterhinch/micropython-iot) which looks great and seems to address any concerns regarding signal drop out. It mostly works ootb too, despite not being authored for windows (select poll is solved with [this gist](https://gist.github.com/gustaebel/b58e887ba93a4d0eba102dae871e11af)).

However I'm not having any luck connecting to my server. I'm connecting to wifi properly and let my server through the firewall, but output via Thonny is:
```
Connection state: True
_readline peer disconnect
reader fail
Connection state: False
Closing sockets.
```
Diving into the client code it seems to suggest getting stuck here:
```
try:
    d = self._sock.readline()
    print(d)
except Exception as e:
    self._verbose and print('_readline exception', d)
    raise
if d == b'':
    self._verbose and print('_readline peer disconnect')
    raise OSError
if d is None:  # Nothing received: wait on server
    if utime.ticks_diff(utime.ticks_ms(), start) > to:
        self._verbose and print('_readline timeout')
        raise OSError
    await asyncio.sleep_ms(0)
else:  # Something received: reset timer
    start = utime.ticks_ms()
    line = b''.join((line, d)) if line else d
```
if I print out the value of d it seems like I get a period where it's None (causing server to wait) but then I'll get a a b'' value which then raises the exception above. Am I missing something fundamental here? I'm using the latest firmwhere [here](https://micropython.org/resources/firmware/RPI_PICO_W-20240105-v1.22.1.uf2) for what it's worth.

Trivial client / server examples using the native socket lib work fine, so I know it's not a hardware / config issue. Which brings me to my second question - it using the iot library absolutely necessary? The page mentions _ESP8266, ESP32 or other target including the Pyboard D._ which would obviously include the Pico W, however it's not explicitly mentioned so just wondering if I'm jumping through these hoops unnecessarily.

Many thanks in advance for your help!

**Comment by f41ardu** [comment]
Hi, 

but, hooo, do you connect to your local network? See [Pico W Documentation on RaspberryPi.org](https://projects.raspberrypi.org/en/projects/get-started-pico-w/2)
as part of my home energy monitoring I've implemented a client(ESP32/2866) to server(Linux) communication to transfer gas meter counts and SML data from my power meter into an rrd database. Both is based on udp and working pretty well. 

Ok, I know many others do this as well. But I love to build my on stuff instead of using AhoiDTU or something else. 

It ii just on my own interest, like yourself I'm now working on a more elaborated version ([REST-API)](https://gist.github.com/f41ardu/ac521a69a90fffa18f733e2e573f201d). 
Far away from being perfect ;-<
If you are interested we can collaborate and we can share the code in a git repository. 
Both can be easily modfoed to transfer any other data created on ESP or PICO PI.
